File: src/fastapi_420/algorithms/sliding_window.py
"""Sliding window rate limiting algorithm."""

import time
from collections import defaultdict
from threading import Lock
from typing import Deque, Dict
from collections import deque

from fastapi_420.algorithms.base import BaseAlgorithm, RateLimitResult


class SlidingWindowAlgorithm(BaseAlgorithm):
    """Sliding window rate limiting.
    
    Uses a sliding time window for more accurate rate limiting.
    Tracks individual request timestamps for precise limiting.
    """
    
    def __init__(self, limit: int, window_seconds: int):
        super().__init__(limit, window_seconds)
        self._requests: Dict[str, Deque[float]] = defaultdict(deque)
        self._lock = Lock()
    
    def _clean_old_requests(self, requests: Deque[float], now: float) -> None:
        """Remove requests older than the window."""
        cutoff = now - self.window_seconds
        while requests and requests[0] < cutoff:
            requests.popleft()
    
    def check_limit(self, key: str, cost: int = 1) -> RateLimitResult:
        now = time.time()
        
        with self._lock:
            requests = self._requests[key]
            self._clean_old_requests(requests, now)
            
            current_count = len(requests)
            
            if current_count + cost > self.limit:
                # Find when the oldest request will expire
                oldest = requests[0] if requests else now
                reset_time = oldest + self.window_seconds
                retry_after = int(reset_time - now) + 1
                return RateLimitResult(
                    allowed=False,
                    limit=self.limit,
                    remaining=0,
                    reset_time=reset_time,
                    retry_after=retry_after
                )
            
            # Add new request timestamps
            for _ in range(cost):
                requests.append(now)
            
            remaining = self.limit - current_count - cost
            # Reset time is when the oldest current request expires
            oldest = requests[0] if requests else now
            reset_time = oldest + self.window_seconds
            
            return RateLimitResult(
                allowed=True,
                limit=self.limit,
                remaining=remaining,
                reset_time=reset_time
            )
    
    def reset(self, key: str) -> None:
        with self._lock:
            if key in self._requests:
                del self._requests[key]
    
    def get_current_usage(self, key: str) -> int:
        now = time.time()
        with self._lock:
            requests = self._requests.get(key, deque())
            self._clean_old_requests(requests, now)
            return len(requests)

File: src/fastapi_420/storage/lua/sliding_window.lua
-- Sliding window rate limiting with Redis
-- KEYS[1] = rate limit key (sorted set)
-- ARGV[1] = limit
-- ARGV[2] = window seconds
-- ARGV[3] = cost
-- ARGV[4] = current timestamp
-- ARGV[5] = request ID (unique)

local key = KEYS[1]
local limit = tonumber(ARGV[1])
local window = tonumber(ARGV[2])
local cost = tonumber(ARGV[3])
local now = tonumber(ARGV[4])
local request_id = ARGV[5]

local cutoff = now - window

-- Remove old entries
redis.call('ZREMRANGEBYSCORE', key, '-inf', cutoff)

-- Count current requests
local current = redis.call('ZCARD', key)

if current + cost > limit then
    local oldest = redis.call('ZRANGE', key, 0, 0, 'WITHSCORES')
    local reset_time = 0
    if #oldest > 0 then
        reset_time = tonumber(oldest[2]) + window
    else
        reset_time = now + window
    end
    local retry_after = math.ceil(reset_time - now)
    return {0, limit, 0, reset_time, retry_after}
end

-- Add new requests
for i = 1, cost do
    redis.call('ZADD', key, now, request_id .. ":" .. i)
end

redis.call('EXPIRE', key, window + 1)

local remaining = limit - current - cost
local oldest = redis.call('ZRANGE', key, 0, 0, 'WITHSCORES')
local reset_time = 0
if #oldest > 0 then
    reset_time = tonumber(oldest[2]) + window
else
    reset_time = now + window
end

return {1, limit, remaining, reset_time, 0}